semanticdb/semanticdb/semanticdb.proto

```diff
@@ -0,0 +1,323 @@
+syntax = "proto3";
+
+package scala.meta.internal.semanticdb;
+
+enum Schema {
+  LEGACY = 0;
+  SEMANTICDB3 = 3;
+  SEMANTICDB4 = 4;
+}
+
+message TextDocuments {
+  repeated TextDocument documents = 1;
+}
+
+message TextDocument {
+  reserved 4, 9;
+  Schema schema = 1;
+  string uri = 2;
+  string text = 3;
+  string md5 = 11;
+  Language language = 10;
+  repeated SymbolInformation symbols = 5;
+  repeated SymbolOccurrence occurrences = 6;
+  repeated Diagnostic diagnostics = 7;
+  repeated Synthetic synthetics = 8;
+}
+
+enum Language {
+  UNKNOWN_LANGUAGE = 0;
+  SCALA = 1;
+  JAVA = 2;
+}
+
+message Range {
+  int32 start_line = 1;
+  int32 start_character = 2;
+  int32 end_line = 3;
+  int32 end_character = 4;
+}
+
+message Location {
+  string uri = 1;
+  Range range = 2;
+}
+
+message Scope {
+  repeated string symlinks = 1;
+  repeated SymbolInformation hardlinks = 2;
+}
+
+message Type {
+  reserved 1, 3, 4, 5, 6, 11, 12, 15, 16;
+  oneof sealed_value {
+    TypeRef typeRef = 2;
+    SingleType singleType = 20;
+    ThisType thisType = 21;
+    SuperType superType = 22;
+    ConstantType constantType = 23;
+    IntersectionType intersectionType = 17;
+    UnionType unionType = 18;
+    WithType withType = 19;
+    StructuralType structuralType = 7;
+    AnnotatedType annotatedType = 8;
+    ExistentialType existentialType = 9;
+    UniversalType universalType = 10;
+    ByNameType byNameType = 13;
+    RepeatedType repeatedType = 14;
+  }
+}
+
+message TypeRef {
+  Type prefix = 1;
+  string symbol = 2;
+  repeated Type type_arguments = 3;
+}
+
+message SingleType {
+  Type prefix = 1;
+  string symbol = 2;
+}
+
+message ThisType {
+  string symbol = 1;
+}
+
+message SuperType {
+  Type prefix = 1;
+  string symbol = 2;
+}
+
+message ConstantType {
+  Constant constant = 1;
+}
+
+message IntersectionType {
+  repeated Type types = 1;
+}
+
+message UnionType {
+  repeated Type types = 1;
+}
+
+message WithType {
+  repeated Type types = 1;
+}
+
+message StructuralType {
+  reserved 1, 2, 3;
+  Type tpe = 4;
+  Scope declarations = 5;
+}
+
+message AnnotatedType {
+  reserved 2;
+  repeated Annotation annotations = 3;
+  Type tpe = 1;
+}
+
+message ExistentialType {
+  reserved 2;
+  Type tpe = 1;
+  Scope declarations = 3;
+}
+
+message UniversalType {
+  reserved 1;
+  Scope type_parameters = 3;
+  Type tpe = 2;
+}
+
+message ByNameType {
+  Type tpe = 1;
+}
+
+message RepeatedType {
+  Type tpe = 1;
+}
+
+message Constant {
+  oneof sealed_value {
+    UnitConstant unitConstant = 1;
+    BooleanConstant booleanConstant = 2;
+    ByteConstant byteConstant = 3;
+    ShortConstant shortConstant = 4;
+    CharConstant charConstant = 5;
+    IntConstant intConstant = 6;
+    LongConstant longConstant = 7;
+    FloatConstant floatConstant = 8;
+    DoubleConstant doubleConstant = 9;
+    StringConstant stringConstant = 10;
+    NullConstant nullConstant = 11;
+  }
+}
```

Comment:
With the new oneof encoding, I can write ADTs in protobuf all day. Thank you so much, @olafurpg and @thesamet!